src/Integration/QualityProfiles/QualityProfileManager.cs

```diff
@@ -0,0 +1,99 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+
+namespace SonarLint.VisualStudio.Integration.RuleSets
+{
+    internal sealed class QualityProfileManager : IDisposable
+    {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly IProjectSystemHelper projectSystemHelper;
+        private readonly IQualityProfileProvider rulesetProvider;
+
+        internal /* for testing purposes */ readonly ConcurrentDictionary<Language, QualityProfile> cachedQualityProfiles =
+            new ConcurrentDictionary<Language, QualityProfile>();
```

Comment:
We should document what the threading model for this class is.

Reply:
I am actually not sure whether we need to handle any concurrency as this class is a singleton.